elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,420 @@
+resource.  Presently the `credential-store` resource contains an `add-alias` operation which can be used to add a credential based on a clear text String,
+the new operations will perform a similar purpose.
+
+The following operations will be added to the `credential-store` resource:
+
+ * `generate-secret-key`
+ * `export-secret-key`
+ * `import-secret-key`
+ 
+Within the management model operations are self describing so making use of dedicated operations makes it easier for tooling to present a meaningful UI to end
+users automatically.
+
+The `credential-store` resource already contains a `remove-alias` operation however this assumes the type of the credential is `PasswordCredential` a new
+attribute will be added to this operation `entry-type` which will default to `PasswordCredential` but will also accept `SecretKeyCredential` as a value
+allowing credentials of type `SecretKeyCredential` to be removed from the credential store.  The `secret-key-credential-store` resource will also have a
+`remove-alias` operation added but as this credential store only supports a single credential type the type parameter will be omitted.
+
+NOTE: The credential store also offers a programmatic API, should the tooling and operations be insufficient for an end user manual population of the credential
+store may remain an option.
+
+===== `generate-secret-key`
+
+The `generate-secret-key` operation will take the following parameters:
+
+ * `alias` - The alias to use when storing the `SecretKeyCredential` in the credential store.
+ * `key-size` - The size of the `AES` key to generate.
+
+In the case of the `credential-store-resource` if the `key-size` attribute is omitted it will default to 256, in the case of the new
+`secret-key-credential-store` resource if omitted it will default to the value defined on the resource which in turn defaults to 256.
+
+===== `export-secret-key`
+
+The `export-secret-key` operation takes a single parameter `alias` which is the alias of the entry to export, the result of the operation is an encoded
+representation of the `SecretKey` suitable for importing elsewhere.
+
+===== `import-secret-key`
+
+The `import-secret-key` operation takes the following parameters:
+
+ * `alias` - The alias to use to store the imported key.
+ * `key` - The encoded representation of the secret key.
+
+=== Expression Resolver Resource
+
+Presently WildFly Core supports an expression resolver that can delegate to a Vault configuration and if that is not available fall back to use either system
+properties or environment variables.  This will be updated to make use of the `CapabilityRegistry` and attempt to lookup a capability using a predefined
+constant `org.wildfly.controller.expression-resolver` which exposes an expression resolver runtime API.  The expression resolver looked up using a capability
+will be used after first attempting to resolve the expression as a Vault expression to resolve any expression.  
```

Comment:
Do I understand it correctly that the new ordering will be: Vault, capabilities, system properties, environment variables?

Reply:
I would need to double check the exact order but 100% this new capaility based resolution is also immediately after Vault.  

Also this should not impact any existing recursion that is (or is not) possible with vault.